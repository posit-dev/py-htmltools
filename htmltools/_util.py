from __future__ import annotations

import hashlib
import importlib
import os
import re
import tempfile
from contextlib import closing
from http.server import SimpleHTTPRequestHandler
from socket import socket
from socketserver import TCPServer
from threading import Thread
from typing import Any, Hashable, Iterable, NamedTuple, Optional, TypeVar, Union

T = TypeVar("T")

HashableT = TypeVar("HashableT", bound=Hashable)

__all__ = ("css",)


def css(collapse_: str = "", **kwargs: str | float | None) -> Optional[str]:
    """
    CSS string helper

    Convenience function for building CSS style declarations (i.e. the string that goes
    into a style attribute, or the parts that go inside curly braces in a full
    stylesheet).

    Parameters
    ----------
    collapse_
        Character to use to collapse properties into a single string; likely "" (the
        default) for style attributes, and either "\n" or None for style blocks.
    **kwargs
        Named style properties, where the name is the property name and the argument is
        the property value.

    Returns
    -------
    :
        A string of CSS style declarations, or ``None`` if no properties were given.

    Example
    -------
    >>> from htmltools import css
    >>> css(font_size = "12px", backgroundColor = "red")
    'font-size:12px;background-color:red;'

    Note
    ----
    CSS uses '-' (minus) as a separator character in property names, which isn't allowed
    in Python's keyword arguments. This function allows you to use '_' (underscore) as a
    separator and/or camelCase notation instead.
    """
    res = ""
    for k, v in kwargs.items():
        if v is None:
            continue
        v = " ".join(v) if isinstance(v, list) else str(v)
        k = re.sub("_", "-", re.sub("([A-Z])", "-\\1", k).lower())
        res += k + ":" + v + ";" + collapse_
    return None if res == "" else res


# Flatten a arbitrarily nested list and remove None. Does not alter input object.
def flatten(x: Iterable[Union[T, None]]) -> list[T]:
    result: list[T] = []
    _flatten_recurse(x, result)
    return result


# Having this separate function and passing along `result` is faster than defining
# a closure inside of `flatten()` (and not passing `result`).
def _flatten_recurse(x: Iterable[T | None], result: list[T]) -> None:
    for item in x:
        if isinstance(item, (list, tuple)):
            # Don't yet know how to specify recursive generic types, so we'll tell
            # the type checker to ignore this line.
            _flatten_recurse(item, result)  # pyright: ignore[reportUnknownArgumentType]
        elif item is not None:
            result.append(item)


# similar to unique() in R (set() doesn't preserve order)
def unique(x: list[HashableT]) -> list[HashableT]:
    # This implementation requires Python 3.7+. Starting with that version, dict
    # order is guaranteed to be the same as insertion order.
    return list(dict.fromkeys(x))


HTML_ESCAPE_TABLE = {
    "&": "&amp;",
    ">": "&gt;",
    "<": "&lt;",
}

HTML_ATTRS_ESCAPE_TABLE = {
    **HTML_ESCAPE_TABLE,
    '"': "&quot;",
    "'": "&apos;",
    "\r": "&#13;",
    "\n": "&#10;",
}


def html_escape(text: str, attr: bool = False) -> str:
    table = HTML_ATTRS_ESCAPE_TABLE if attr else HTML_ESCAPE_TABLE
    if not re.search("|".join(table), text):
        return text
    for key, value in table.items():
        text = text.replace(key, value)
    return text


# Backwards compatibility with faicons 0.2.1
_html_escape = html_escape


# similar to base::system.file()
def package_dir(package: str) -> str:
    with tempfile.TemporaryDirectory():
        pkg_file = importlib.import_module(".", package=package).__file__
        if pkg_file is None:
            raise ImportError(f"Couldn't load package {package}")
        return os.path.dirname(pkg_file)


# Backwards compatibility with shinywidgets 0.1.4
_package_dir = package_dir


def hash_deterministic(s: str) -> str:
    """
    Returns a deterministic hash of the given string.
    """
    return hashlib.sha1(s.encode("utf-8")).hexdigest()


class _HttpServerInfo(NamedTuple):
    port: int
    thread: Thread


_http_servers: dict[str, _HttpServerInfo] = {}


def ensure_http_server(path: str) -> int:
    server = _http_servers.get(path)
    if server:
        return server.port

    _http_servers[path] = start_http_server(path)
    return _http_servers[path].port


def start_http_server(path: str) -> _HttpServerInfo:
    port: int = get_open_port()
    th: Thread = Thread(target=http_server, args=(port, path), daemon=True)
    th.start()
    return _HttpServerInfo(port=port, thread=th)


def http_server(port: int, path: str):
    class Handler(SimpleHTTPRequestHandler):
        def __init__(self, *args: Any, **kwargs: Any):
            super().__init__(*args, directory=path, **kwargs)

        def log_message(
            self, format, *args  # pyright: ignore[reportMissingParameterType]
        ):
            pass

    with TCPServer(("", port), Handler) as httpd:
        httpd.serve_forever()


def get_open_port() -> int:
    with closing(socket()) as sock:
        sock.bind(("", 0))
        return sock.getsockname()[1]
