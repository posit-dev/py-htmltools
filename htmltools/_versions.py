versions = {"react":"18.0.0","react-dom":"18.0.0","html-react-parser":"1.4.10"}