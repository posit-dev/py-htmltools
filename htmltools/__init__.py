__version__ = "0.4.0"

from . import svg, tags
from ._core import TagAttrArg  # pyright: ignore[reportUnusedImport]  # noqa: F401
from ._core import TagChildArg  # pyright: ignore[reportUnusedImport]  # noqa: F401
from ._core import (
    HTML,
    HTMLDependency,
    HTMLDocument,
    HTMLTextDocument,
    MetadataNode,
    RenderedHTML,
    Tag,
    TagAttrs,
    TagAttrValue,
    TagChild,
    TagFunction,
    Tagifiable,
    TagList,
    TagNode,
    head_content,
)
from ._util import css, html_escape
from .tags import (
    a,
    br,
    code,
    div,
    em,
    h1,
    h2,
    h3,
    h4,
    h5,
    h6,
    hr,
    img,
    p,
    pre,
    span,
    strong,
)

__all__ = (
    "svg",
    "tags",
    "HTML",
    "HTMLDependency",
    "HTMLDocument",
    "HTMLTextDocument",
    "MetadataNode",
    "RenderedHTML",
    "Tag",
    "TagAttrs",
    "TagAttrValue",
    "TagChild",
    "TagFunction",
    "Tagifiable",
    "TagList",
    "TagNode",
    "head_content",
    "css",
    "html_escape",
    "a",
    "br",
    "code",
    "div",
    "em",
    "h1",
    "h2",
    "h3",
    "h4",
    "h5",
    "h6",
    "hr",
    "img",
    "p",
    "pre",
    "span",
    "strong",
)


import typing as _typing

# Setting this will control how HTML dependencies are rendered. Normally they are not
# visible, but if set to "json", they will be serialized as JSON in a <script> tag.
html_dependency_render_mode: _typing.Literal["json", "invisible"] = "invisible"
