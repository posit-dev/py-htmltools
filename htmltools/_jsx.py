"""
WARNING: this is a highly experimental/unsupported/private module for JSX/React
components. Although JSXTag currently allows for HTML tags on attributes and children,
that's an issue for HTML() and <script> tags, so using normal HTML tags inside JSX
components may become unsupported in a future version (see #26 and #28)
"""
from __future__ import annotations

import copy
import re
from typing import Any, Callable, Dict, Iterable, Mapping, Optional, cast

from ._core import (
    HTML,
    HTMLDependency,
    MetadataNode,
    ScriptItem,
    Tag,
    TagAttrValue,
    Tagifiable,
    TagList,
    TagNode,
)
from ._versions import versions

# CPS 6/8/2022: this module is currently too experimental to be recommended for use.
# It's not clear that it's worth the effort to support it, and if we do, we may want to
# remove the ability to nest HTML() and <script> tags inside JSXTag/JSXTagAttrValue
__all__ = (
    # "jsx",
    # "jsx_tag_create",
    # "JSXTag",
    # "JSXTagAttrValue",
)

# A JSX tag attribute can be anything.
JSXTagAttrValue = object


class JSXTagAttrDict(Dict[str, JSXTagAttrValue]):
    def __init__(self, **kwargs: JSXTagAttrValue) -> None:
        super().__init__()
        self.update(**kwargs)

    def __setitem__(self, name: str, value: JSXTagAttrValue) -> None:
        nm = self._normalize_attr_name(name)
        super().__setitem__(nm, value)

    def update(  # type: ignore[reportIncompatibleMethodOverride] # TODO-future: fix typing
        self,
        *args: Mapping[str, JSXTagAttrValue],
        **kwargs: JSXTagAttrValue,
    ) -> None:
        for arg in args:
            self._update(arg)
        self._update(kwargs)

    def _update(self, __m: Mapping[str, JSXTagAttrValue]) -> None:
        attrs: dict[str, JSXTagAttrValue] = {}
        for key, val in __m.items():
            attrs[self._normalize_attr_name(key)] = val
        super().update(**attrs)

    @staticmethod
    def _normalize_attr_name(x: str) -> str:
        if x.endswith("_"):
            x = x[:-1]
        return x.replace("_", "-")


class JSXTag:
    """
    Create a JSX tag.

    Warning
    -------
    This class shouldn't be used directly to create a JSX tag. Instead, use the
    jsx_tag_create() function.

    See Also
    --------
    jsx_tag_create
    """

    def __init__(
        self,
        _name: str,
        *args: TagNode,
        allowedProps: Optional[list[str]] = None,
        **kwargs: JSXTagAttrValue,
    ) -> None:
        pieces = _name.split(".")
        if pieces[-1][:1] != pieces[-1][:1].upper():
            raise NotImplementedError("JSX tags must be start with a capital letter.")

        if allowedProps:
            for k in kwargs.keys():
                if k not in allowedProps:
                    raise NotImplementedError(f"{k} is not a valid prop for {_name}")

        self.name: str = _name
        # Unlike HTML tags, JSX tag attributes can be anything.
        self.attrs: JSXTagAttrDict = JSXTagAttrDict(**kwargs)
        self.children: TagList = TagList(*args)

    def extend(self, x: Iterable[TagNode]) -> None:
        self.children.extend(x)

    def append(self, *args: TagNode) -> None:
        self.children.append(*args)

    def tagify(self) -> Tag:
        metadata_nodes: list[MetadataNode] = []

        # This function is recursively applied to the attributes and children. It does
        # two things: For attrs/children that are Tagifiable but NOT Tags or JSXTags, it
        # calls .tagify(), so that we have a fully tagified tree. Then it collects any
        # metadata nodes. This could be done in two separate passes, but it's more
        # efficient to do it in one pass.
        def tagify_tagifiable_and_get_metadata(x: Any) -> Any:
            if isinstance(x, Tagifiable) and not isinstance(x, (Tag, JSXTag)):
                x = x.tagify()
            else:
                x = copy.copy(x)
            if isinstance(x, MetadataNode):
                metadata_nodes.append(x)
            return x

        cp = copy.copy(self)
        _walk_attrs_and_children(cp, tagify_tagifiable_and_get_metadata)

        # When _render_react_js()  is called on a JSXTag object, we'll recurse, but
        # instead of calling the standard Tag.get_html_string() method to format the
        # object, we'll recurse using _render_react_js(), which descends into the tree
        # and formats objects appropriately for inside of a JSX element.
        component = _render_react_js(self, 2, "\n")

        # Ideally, we'd use document.currentScript.after() to insert the component
        # directly after the script tag, but when dynamically rendered via jQuery (i.e.,
        # @render_ui()), the script tag actually executes in <head>, which makes the
        # document.currentScript reference basically useless for this purpose. So,
        # instead, to make this work when dynamically rendered, this script queries for
        # the first JSXTag script that hasn't yet rendered, and insert the component
        # there. It seems like this should work fine as long as these script tags are
        # synchronously executed in order, but if we ever need them to render
        # asynchronously, it might make sense to use a unique ID for each script tag
        # instead.
        js = "\n".join(
            [
                "(function() {",
                "  var container = new DocumentFragment();",
                "  ReactDOM.render(",
                component,
                "  , container);",
                "  var thisScript = document.querySelector('script[data-needs-render]');",
                f"  if (!thisScript) throw new Error('Failed to render JSXTag(\"{self.name}\")');",
                "  thisScript.after(container);",
                "  thisScript.removeAttribute('data-needs-render');",
                "})();",
            ]
        )

        return Tag(
            "script",
            {
                "type": "text/javascript",
                "data_needs_render": True,
            },
            HTML("\n" + js + "\n"),
            _lib_dependency("react", script={"src": "react.production.min.js"}),
            _lib_dependency("react-dom", script={"src": "react-dom.production.min.js"}),
            *metadata_nodes,
        )

    def __str__(self) -> str:
        return str(self.tagify())

    def __repr__(self) -> str:
        return self.__str__()

    def _repr_html_(self) -> str:
        return self.__str__()


def _walk_attrs_and_children(x: Any, fn: Callable[[Any], Any]) -> Any:
    x = fn(x)

    if isinstance(x, Tag):
        for i, child in enumerate(x.children):
            x.children[i] = _walk_attrs_and_children(child, fn)
    elif isinstance(x, JSXTag):
        for key, value in x.attrs.items():
            x.attrs[key] = _walk_attrs_and_children(value, fn)
        for i, child in enumerate(x.children):
            x.children[i] = _walk_attrs_and_children(child, fn)
    elif isinstance(x, Tagifiable):
        # Don't do anything here?
        pass

    return x


# Return a string representing the rendered HTML for the given JSXTag object. The
# metadata_nodes object collects MetadataNode objects in the tree, and is altered by
# reference as a side-effect of this function.
def _render_react_js(x: TagNode, indent: int, eol: str) -> str:
    indent_str = "  " * indent

    if isinstance(x, MetadataNode):
        return ""
    elif isinstance(x, str):
        return indent_str + '"' + x.replace('"', '\\"') + '"'

    if isinstance(x, JSXTag):
        nm = x.name
    elif isinstance(x, Tag):
        nm = "'" + x.name + "'"
    else:
        raise TypeError("x must be a tag or JSXTag object. Did you run tagify()?")

    res: str = indent_str + "React.createElement("

    if len(x.attrs) == 0 and len(x.children) == 0:
        return res + nm + ")"

    res += f"{eol}{indent_str}  {nm}, "

    res += "{"
    is_first_attr = True
    for k, v in x.attrs.items():
        if not is_first_attr:
            res += ", "
        is_first_attr = False
        if k == "style":
            v = _serialize_style_attr(v)
        else:
            v = _serialize_attr(v)
        res += f'"{k}": {v}'
    res += "}"

    if len(x.children) == 0:
        return res + ")"

    for child in x.children:
        child_str = _render_react_js(child, indent + 1, eol)
        if child_str != "":
            res += "," + eol + child_str

    return res + eol + indent_str + ")"


# Unfortunately we can't use json.dumps() here because I don't know how to
# avoid quoting jsx(), jsx_tag(), tag(), etc.
def _serialize_attr(x: object) -> str:
    if x is None:
        return "null"
    if isinstance(x, Tag) or isinstance(x, JSXTag):
        return _render_react_js(x, 0, "\n")
    if isinstance(x, (list, tuple)):
        return (
            "[" + ", ".join([_serialize_attr(y) for y in cast(Iterable[Any], x)]) + "]"
        )
    if isinstance(x, dict):
        x1: dict[str, Any] = x  # make pyright happy
        return "{" + ", ".join([f'"{y}": ' + _serialize_attr(x1[y]) for y in x1]) + "}"
    if isinstance(x, bool):
        return str(x).lower()
    if isinstance(x, (jsx, int, float)):
        return str(x)
    return '"' + str(x).replace('"', '\\"') + '"'


def _serialize_style_attr(x: object) -> str:
    if x is None:
        return "{}"
    # Parse CSS strings into a dict (because React requires it)
    if isinstance(x, str):
        x = cast(
            "list[tuple[str, str]]",
            [tuple(y.split(":")) for y in x.split(";") if re.search(":", y)],
        )
        x = dict(x)
    if isinstance(x, dict):
        return _serialize_attr(cast(Dict[str, Any], x))
    else:
        raise TypeError("The style attribute must be a dict() or string.")


def jsx_tag_create(
    name: str, allowedProps: Optional[list[str]] = None
) -> Callable[..., JSXTag]:
    """
    Create a function that creates a JSXTag object.

    Parameters
    ----------
    name
        The name of the JSX tag (should be camelCase and start with a capital letter).
    allowedProps
        A list of allowed properties for the tag. If ``None``, all properties are
        allowed.

    Returns
    -------
    :
        JSXTag

    Examples
    --------
    >>> from htmltools import jsx_tag_create
    >>> MyTag = jsx_tag_create("MyTag")
    >>> MyTag(id="foo", class_="bar")
    <script type="text/javascript">
    (function() {
      var container = new DocumentFragment();
      ReactDOM.render(
        React.createElement(
          MyTag, {"id": "foo", "class": "bar"})
      , container);
      document.currentScript.after(container);
    })();
    </script>
    """

    def create_tag(
        *args: TagNode,
        **kwargs: TagAttrValue,
    ) -> JSXTag:
        return JSXTag(name, *args, allowedProps=allowedProps, **kwargs)

    create_tag.__name__ = name

    return create_tag


# --------------------------------------------------------
# jsx expressions
# --------------------------------------------------------
class jsx(str):
    """
    Mark a string as a JSX expression.

    Example
    -------
    >>> Foo = JSXTag("Foo")
    >>> Foo(prop = "A string", jsxProp = jsx("() => console.log('here')"))
    <script type="text/javascript">
    (function() {
      var container = new DocumentFragment();
      ReactDOM.render(
        React.createElement(
          Foo, {"prop": "A string", "jsxProp": () => console.log('here')})
      , container);
      document.currentScript.after(container);
    })();
    </script>
    """

    def __new__(cls, *args: str) -> "jsx":
        return super().__new__(  # pyright: ignore[reportGeneralTypeIssues]
            cls, "\n".join(args)
        )

    # jsx() + jsx() should return jsx()
    def __add__(self, other: "str | jsx") -> str:
        res = str.__add__(self, other)
        return jsx(res) if isinstance(other, jsx) else res


def _lib_dependency(pkg: str, script: ScriptItem) -> HTMLDependency:
    return HTMLDependency(
        name=pkg,
        version=versions[pkg],
        source={"package": "htmltools", "subdir": "lib/" + pkg},
        script=script,
    )
