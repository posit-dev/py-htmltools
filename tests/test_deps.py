import tempfile
import textwrap
from pathlib import Path

from htmltools import HTMLDependency, HTMLDocument, TagList, div, tags


def test_init_errors():
    HTMLDependency("a", "1.1", source=None)

    try:
        HTMLDependency("a", "1.1", source=4)  # type: ignore
    except TypeError as e:
        assert "to be a dict" in str(e)

    try:
        HTMLDependency("a", "1.1", source={"not": "valid"})  # type: ignore
    except TypeError as e:
        assert "to have either" in str(e)


def test_dep_resolution():
    a1_1 = HTMLDependency("a", "1.1", source={"subdir": "foo"}, script={"src": "a1.js"})
    a1_2 = HTMLDependency(
        "a", "1.2", source={"href": "http://a.test.com"}, script={"src": "a2.js"}
    )
    a1_2_1 = HTMLDependency(
        "a", "1.2.1", source={"subdir": "foo"}, script={"src": "a3.js"}
    )
    b1_9 = HTMLDependency("b", "1.9", source={"subdir": "foo"}, script={"src": "b1.js"})
    b1_10 = HTMLDependency(
        "b", "1.10", source={"href": "http://b.test.com"}, script={"src": "b2.js"}
    )
    c1_0 = HTMLDependency("c", "1.0", source={"subdir": "foo"}, script={"src": "c1.js"})
    test = TagList(a1_1, b1_9, b1_10, a1_2, a1_2_1, b1_9, b1_10, c1_0)
    assert HTMLDocument(test).render(lib_prefix=None)["html"] == textwrap.dedent(
        """\
        <!DOCTYPE html>
        <html>
          <head>
            <meta charset="utf-8"/>
            <script type="application/html-dependencies">a[1.2.1];b[1.10];c[1.0]</script>
            <script src="a-1.2.1/a3.js"></script>
            <script src="http://b.test.com/b2.js"></script>
            <script src="c-1.0/c1.js"></script>
          </head>
          <body></body>
        </html>"""
    )

    assert HTMLDocument(test).render(lib_prefix="libfoo")["html"] == textwrap.dedent(
        """\
        <!DOCTYPE html>
        <html>
          <head>
            <meta charset="utf-8"/>
            <script type="application/html-dependencies">a[1.2.1];b[1.10];c[1.0]</script>
            <script src="libfoo/a-1.2.1/a3.js"></script>
            <script src="http://b.test.com/b2.js"></script>
            <script src="libfoo/c-1.0/c1.js"></script>
          </head>
          <body></body>
        </html>"""
    )


# Test out renderTags and findDependencies when tags are inline
def test_inline_deps(snapshot):
    a1_1 = HTMLDependency("a", "1.1", source={"subdir": "foo"}, script={"src": "a1.js"})
    a1_2 = HTMLDependency("a", "1.2", source={"subdir": "foo"}, script={"src": "a2.js"})
    tests = [
        TagList(a1_1, div("foo"), "bar"),
        TagList(a1_1, div("foo"), a1_2, "bar"),
        div(a1_1, div("foo"), "bar"),
        TagList([a1_1, div("foo")], "bar"),
        div([a1_1, div("foo")], "bar"),
    ]
    html_ = "\n\n".join(
        [HTMLDocument(t).render(lib_prefix=None)["html"] for t in tests]
    )
    assert html_ == snapshot


def test_append_deps():
    a1_1 = HTMLDependency("a", "1.1", source={"subdir": "foo"}, script={"src": "a1.js"})
    a1_2 = HTMLDependency(
        "a", "1.2", source={"href": "http://foo.com"}, script={"src": "a2.js"}
    )
    b1_0 = HTMLDependency("b", "1.0", source={"subdir": "foo"}, script={"src": "b1.js"})

    expected_result = textwrap.dedent(
        """\
        <!DOCTYPE html>
        <html>
          <head>
            <meta charset="utf-8"/>
            <script type="application/html-dependencies">a[1.2];b[1.0]</script>
            <script src="http://foo.com/a2.js"></script>
            <script src="b-1.0/b1.js"></script>
          </head>
          <body>
            <div></div>
          </body>
        </html>"""
    )

    x = div(a1_1, b1_0)
    x.append(a1_2)
    assert HTMLDocument(x).render(lib_prefix=None)["html"] == expected_result

    y = div(a1_1)
    y.append([a1_2, b1_0])
    assert HTMLDocument(y).render(lib_prefix=None)["html"] == expected_result

    z = div()
    z.append([a1_1, b1_0])
    z.append(a1_2)
    assert HTMLDocument(z).render(lib_prefix=None)["html"] == expected_result


def test_script_input():
    def fake_dep(**kwargs):
        return HTMLDependency(
            "a", "1.0", source={"package": None, "subdir": "srcpath"}, **kwargs
        )

    dep1 = fake_dep(
        script={"src": "js/foo bar.js"}, stylesheet={"href": "css/bar foo.css"}
    )
    dep2 = fake_dep(
        script=[{"src": "js/foo bar.js"}], stylesheet=[{"href": "css/bar foo.css"}]
    )
    assert dep1 == dep2
    # Make sure repeated calls to as_html() repeatedly encode
    test = TagList([dep1, dep2])
    for _ in range(2):
        assert HTMLDocument(test).render()["html"] == textwrap.dedent(
            """\
            <!DOCTYPE html>
            <html>
              <head>
                <meta charset="utf-8"/>
                <script type="application/html-dependencies">a[1.0]</script>
                <link href="lib/a-1.0/css/bar%20foo.css" rel="stylesheet"/>
                <script src="lib/a-1.0/js/foo%20bar.js"></script>
              </head>
              <body></body>
            </html>"""
        )


def test_head_output():
    a = HTMLDependency(
        "a",
        "1.0",
        source={"subdir": "foo"},
        head=tags.script("1 && 1"),
    )
    assert a.as_html_tags().get_html_string() == "<script>1 && 1</script>"

    b = HTMLDependency(
        "a",
        "1.0",
        source={"subdir": "foo"},
        head="<script>1 && 1</script>",
    )
    assert b.as_html_tags().get_html_string() == "<script>1 && 1</script>"

    assert (
        b.serialize_to_script_json(indent=4).get_html_string()
        == """<script type="application/json" data-html-dependency="">{
    "name": "a",
    "version": "1.0",
    "source": {
        "subdir": "foo"
    },
    "script": [],
    "stylesheet": [],
    "meta": [],
    "all_files": false,
    "head": "<script>1 && 1<\\/script>"
}</script>"""
    )


def test_meta_output():
    a = HTMLDependency(
        "a",
        "1.0",
        source={"subdir": "foo"},
        script={"src": "a1.js"},
        meta={"name": "viewport", "content": "width=device-width, initial-scale=1"},
    )

    b = HTMLDependency(
        "b",
        "2.0",
        source={"subdir": "foo"},
        meta=[{"name": "x", "content": "x-value"}, {"name": "y", "content": "y-value"}],
    )

    assert str(a.as_html_tags()) == textwrap.dedent(
        """\
        <meta name="viewport" content="width=device-width, initial-scale=1"/>
        <script src="lib/a-1.0/a1.js"></script>"""
    )
    assert str(b.as_html_tags()) == textwrap.dedent(
        """\
        <meta name="x" content="x-value"/>
        <meta name="y" content="y-value"/>"""
    )

    # Combine the two in an HTMLDocument and render; all meta tags should show up.
    combined_html = HTMLDocument(TagList(a, b)).render()["html"]
    assert (
        combined_html.find(
            '<meta name="viewport" content="width=device-width, initial-scale=1"/>'
        )
        != -1
    )
    assert combined_html.find('<meta name="x" content="x-value"/>') != -1
    assert combined_html.find('<meta name="y" content="y-value"/>') != -1


def test_source_dep_as_dict():
    a = HTMLDependency(
        "a",
        "1.0",
        source={"package": "htmltools", "subdir": "bar"},
        script={"src": "a1.js"},
        meta={"name": "viewport", "content": "width=device-width, initial-scale=1"},
    )
    assert a.as_dict() == {
        "name": "a",
        "version": "1.0",
        "script": [{"src": "lib/a-1.0/a1.js"}],
        "stylesheet": [],
        "meta": [
            {"name": "viewport", "content": "width=device-width, initial-scale=1"}
        ],
        "head": None,
    }

    b = HTMLDependency(
        "b",
        "2.0",
        source={"package": "htmltools", "subdir": "bar"},
        stylesheet=[{"href": "b1.css"}, {"href": "b2.css"}],
        head=tags.script("1 && 1"),
    )
    assert b.as_dict(lib_prefix=None) == {
        "name": "b",
        "version": "2.0",
        "script": [],
        "stylesheet": [
            {"href": "b-2.0/b1.css", "rel": "stylesheet"},
            {"href": "b-2.0/b2.css", "rel": "stylesheet"},
        ],
        "meta": [],
        "head": "<script>1 && 1</script>",
    }


def test_url_dep_as_dict():
    u = HTMLDependency(
        "u",
        "1.1",
        source={"href": "https://example.com"},
        script={"src": "u1.js"},
        stylesheet={"href": "u1.css"},
        meta={"name": "viewport", "content": "width=device-width, initial-scale=1"},
    )
    assert u.as_dict() == {
        "name": "u",
        "version": "1.1",
        "script": [{"src": "https://example.com/u1.js"}],
        "stylesheet": [{"href": "https://example.com/u1.css", "rel": "stylesheet"}],
        "meta": [
            {"name": "viewport", "content": "width=device-width, initial-scale=1"}
        ],
        "head": None,
    }

    v = HTMLDependency(
        "v",
        "2.1",
        source={"href": "https://test.com/subdir/trailing/slash/"},
        stylesheet=[{"href": "v1.css"}, {"href": "v2.css"}],
        head=tags.script("1 && 1"),
    )
    assert v.as_dict(lib_prefix=None) == {
        "name": "v",
        "version": "2.1",
        "script": [],
        "stylesheet": [
            {
                "href": "https://test.com/subdir/trailing/slash/v1.css",
                "rel": "stylesheet",
            },
            {
                "href": "https://test.com/subdir/trailing/slash/v2.css",
                "rel": "stylesheet",
            },
        ],
        "meta": [],
        "head": "<script>1 && 1</script>",
    }


def test_copy_to():
    with tempfile.TemporaryDirectory() as tmpdir1:
        dep1 = HTMLDependency(
            "w",
            "1.0",
            source={"package": "htmltools", "subdir": "libtest/testdep"},
            all_files=True,
        )
        dep1.copy_to(tmpdir1)
        assert (Path(tmpdir1) / "w-1.0" / "testdep.css").exists()
        assert (Path(tmpdir1) / "w-1.0" / "testdep.js").exists()

    with tempfile.TemporaryDirectory() as tmpdir2:
        dep2 = HTMLDependency(
            "w",
            "1.0",
            source={"package": "htmltools", "subdir": "libtest/testdep"},
            all_files=False,
        )
        dep2.copy_to(tmpdir2)
        assert not (Path(tmpdir2) / "w-1.0" / "testdep.css").exists()
        assert not (Path(tmpdir2) / "w-1.0" / "testdep.js").exists()

    with tempfile.TemporaryDirectory() as tmpdir3:
        dep3 = HTMLDependency(
            "w",
            "1.0",
            source={"package": "htmltools", "subdir": "libtest/testdep"},
            stylesheet={"href": "testdep.css"},
            all_files=False,
        )
        dep3.copy_to(tmpdir3)
        assert (Path(tmpdir3) / "w-1.0" / "testdep.css").exists()
        assert not (Path(tmpdir3) / "w-1.0" / "testdep.js").exists()

    with tempfile.TemporaryDirectory() as tmpdir4:
        dep4 = HTMLDependency(
            "w",
            "1.0",
            source={"package": "htmltools", "subdir": "libtest/testdep"},
            stylesheet={"href": "testdep.css"},
            script={"src": "testdep.js"},
            all_files=False,
        )
        dep4.copy_to(tmpdir4)
        assert (Path(tmpdir4) / "w-1.0" / "testdep.css").exists()
        assert (Path(tmpdir4) / "w-1.0" / "testdep.js").exists()
