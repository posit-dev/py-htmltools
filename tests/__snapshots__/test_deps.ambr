# serializer version: 1
# name: test_inline_deps
  '''
  <!DOCTYPE html>
  <html>
    <head>
      <meta charset="utf-8"/>
      <script type="application/html-dependencies">a[1.1]</script>
      <script src="a-1.1/a1.js"></script>
    </head>
    <body>
      <div>foo</div>
      bar
    </body>
  </html>
  
  <!DOCTYPE html>
  <html>
    <head>
      <meta charset="utf-8"/>
      <script type="application/html-dependencies">a[1.2]</script>
      <script src="a-1.2/a2.js"></script>
    </head>
    <body>
      <div>foo</div>
      bar
    </body>
  </html>
  
  <!DOCTYPE html>
  <html>
    <head>
      <meta charset="utf-8"/>
      <script type="application/html-dependencies">a[1.1]</script>
      <script src="a-1.1/a1.js"></script>
    </head>
    <body>
      <div>
        <div>foo</div>
        bar
      </div>
    </body>
  </html>
  
  <!DOCTYPE html>
  <html>
    <head>
      <meta charset="utf-8"/>
      <script type="application/html-dependencies">a[1.1]</script>
      <script src="a-1.1/a1.js"></script>
    </head>
    <body>
      <div>foo</div>
      bar
    </body>
  </html>
  
  <!DOCTYPE html>
  <html>
    <head>
      <meta charset="utf-8"/>
      <script type="application/html-dependencies">a[1.1]</script>
      <script src="a-1.1/a1.js"></script>
    </head>
    <body>
      <div>
        <div>foo</div>
        bar
      </div>
    </body>
  </html>
  '''
# ---
